flocker/acceptance/endtoend/test_diagnostics.py

```diff
@@ -0,0 +1,103 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Tests for the flocker-diagnostics.
+"""
+
+import os
+import tarfile
+
+from twisted.internet import reactor
+from twisted.python.filepath import FilePath
+from twisted.trial.unittest import TestCase
+
+from ...common.runner import run_ssh, download_file
+from ..testtools import require_cluster
+
+
+class DiagnosticsTests(TestCase):
+    """
+    Tests for ``flocker-diagnostics``.
+    """
+    @require_cluster(1)
+    def test_export(self, cluster):
+        """
+        ``flocker-diagnostics`` creates an archive of all Flocker service logs
+        and server diagnostics information.
+        """
+        node_address = cluster.control_node.public_address
+
+        def create_archive():
+            remote_archive_path = []
+            return run_ssh(
+                reactor,
+                'root',
+                node_address,
+                ['flocker-diagnostics'],
+                handle_stdout=remote_archive_path.append
+            ).addCallback(
+                lambda ignored: FilePath(remote_archive_path[0])
+            )
+        creating = create_archive()
+
+        def download_archive(remote_archive_path):
+            local_archive_path = FilePath(self.mktemp())
+            return download_file(
+                reactor,
+                'root',
+                node_address,
+                remote_archive_path,
+                local_archive_path,
+            ).addCallback(lambda ignored: local_archive_path)
+        downloading = creating.addCallback(download_archive)
+
+        def verify_archive(local_archive_path):
+            with tarfile.open(local_archive_path.path) as f:
+                actual_basenames = set()
+                for name in f.getnames():
+                    basename = os.path.basename(name)
+                    if name == basename:
+                        # Ignore the directory entry
+                        continue
+                    actual_basenames.add(basename)
+
+            expected_basenames = set([
+                'flocker-control_startup.gz',
+                'flocker-control_eliot.gz',
+                'flocker-dataset-agent_startup.gz',
+                'flocker-dataset-agent_eliot.gz',
+                'flocker-container-agent_startup.gz',
+                'flocker-container-agent_eliot.gz',
+                'flocker-version',
+                'docker-info',
+                'docker-version',
+                'os-release',
+                'syslog.gz',
+                'uname',
+                'service-status',
+                'ip-addr',
+                'hostname',
+            ])
```

Comment:
Github makes it difficult to tell, but I think you asked me what my honest opinion was about this test. 

It tests a large swathe of code and makes a relatively imprecise assertion. I would generally try to avoid tests like this. However, sometimes they are the only way (or the only cost-feasible way) of getting coverage. Because Python is so easy to mess up, coverage is important. 

I guess I would prefer direct tests of the `FlockerDebugArchive.create` method. The actual script execution stuff is presumably well tested by the standard script tests, and the content of the script is so simple that the value of an integration test there is low. It's possible that direct tests of `create` would require so much set up that they too would end up being [change detector tests](http://googletesting.blogspot.co.uk/2015/01/testing-on-toilet-change-detector-tests.html). I couldn't say without trying.

Fundamentally it's a messy problem, and I think this test is OK. Does that help?